src/lib.rs

```diff
@@ -237,38 +189,337 @@ unsafe fn new_array<A: Array>() -> A {
+    /// buf.push_front(1);
+    ///
+    /// let overflow = buf.push_front(2);
+    ///
+    /// assert_eq!(buf.back(), Some(&0));
+    /// assert_eq!(overflow, Err(CapacityError { element: 2 }));
+    /// ```
+    pub fn push_front(&mut self, element: A::Item) {
+        if self.is_full() {
+            self.pop_back();
+        } 
+
+        unsafe { self.push_front_unchecked(element); }        
+    }
+
+    /// Appends an element to the back of a buffer
+    ///
+    /// Return `None` if the push succeeds, or and return `Some(` *element* `)`
+    /// if the vector is full.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use arraydeque::ArrayDeque;
+    /// use arraydeque::CapacityError;
+    ///
+    /// let mut buf: ArrayDeque<[_; 3]> = ArrayDeque::new();
+    ///
+    /// buf.push_back(0);
+    /// buf.push_back(1);
+    ///
+    /// let overflow = buf.push_back(2);
+    ///
+    /// assert_eq!(buf.back(), Some(&1));
+    /// assert_eq!(overflow, Err(CapacityError { element: 2 }));
+    /// ```
+    pub fn push_back(&mut self, element: A::Item) {
+        if self.is_full() {
+            self.pop_front();
+        } 
+
+        unsafe { self.push_back_unchecked(element); }   
+    }
+
+    /// Inserts an element at `index` within the `ArrayDeque`. Whichever
+    /// end is closer to the insertion point will be moved to make room,
+    /// and all the affected elements will be moved to new positions.
+    ///
+    /// Return `None` if the push succeeds, or and return `Some(` *element* `)`
+    /// if the vector is full.
+    ///
+    /// Element at index 0 is the front of the queue.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `index` is greater than `ArrayDeque`'s length
+    ///
+    /// # Examples
+    /// ```
+    /// use arraydeque::ArrayDeque;
+    /// use arraydeque::CapacityError;
+    ///
+    /// let mut buf: ArrayDeque<[_; 4]> = ArrayDeque::new();
+    ///
+    /// buf.push_back(0);
+    /// buf.push_back(1);
+    ///
+    /// buf.insert(0, 2);
+    ///
+    /// let overflow = buf.insert(1, 3);
+    ///
+    /// assert_eq!(buf[0], 2);
+    /// assert_eq!(overflow, Err(CapacityError { element: 3 }));
+    /// ```
+    pub fn insert(&mut self, index: usize, element: A::Item) {
+        assert!(index <= self.len(), "index out of bounds");
 
-        // NoDrop inhibits array's drop
-        // panic safety: NoDrop::drop will trigger on panic, so the inner
-        // array will not drop even after panic.
+        if self.is_full() {
+            self.pop_front();
+        }
+
+        unsafe { self.insert_unchecked(index, element); }
+    }
+
+    /// Moves all the elements of `other` into `Self`, leaving `other` empty.
+    ///
+    /// Does not extract more items than there is space for. No error
+    /// occurs if there are more iterator elements.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use arraydeque::ArrayDeque;
+    ///
+    /// let mut buf: ArrayDeque<[_; 8]> = ArrayDeque::new();
```

Comment:
You're not creating a `Wrapping` deque here.

Reply:
Yup, not 'document for wrapping behavior' yet. XD